compiler/noirc_frontend/src/parser/parser.rs

```diff
@@ -91,1764 +44,504 @@ use traits::where_clause;
-    let array_rhs = expr_parser
-        .clone()
-        .delimited_by(just(Token::LeftBracket), just(Token::RightBracket))
-        .map(UnaryRhs::ArrayIndex);
-
-    // `as Type` in `atom as Type`
-    let cast_rhs = keyword(Keyword::As)
-        .ignore_then(type_parser.clone())
-        .map(UnaryRhs::Cast)
-        .labelled(ParsingRuleLabel::Cast);
-
-    // A turbofish operator is optional in a method call to specify generic types
-    let turbofish = primitives::turbofish(type_parser);
-
-    // `::<A, B>!(arg1, .., argN)` with the turbofish and macro portions being optional.
-    let method_call_rhs = turbofish
-        .then(just(Token::Bang).or_not())
-        .then(parenthesized(expression_list(expr_parser.clone())))
-        .validate(|((turbofish, macro_call), args), span, emit| {
-            if turbofish.as_ref().map_or(false, |generics| !generics.named_args.is_empty()) {
-                let reason = ParserErrorReason::AssociatedTypesNotAllowedInMethodCalls;
-                emit(ParserError::with_reason(reason, span));
+    fn eat_ident(&mut self) -> Option<Ident> {
+        if let Some(token) = self.eat_kind(TokenKind::Ident) {
+            match token.into_token() {
+                Token::Ident(ident) => Some(Ident::new(ident, self.previous_token_span)),
+                _ => unreachable!(),
             }
+        } else {
+            None
+        }
+    }
 
-            let macro_call = macro_call.is_some();
-            let turbofish = turbofish.map(|generics| generics.ordered_args);
-            UnaryRhsMethodCall { turbofish, macro_call, args }
-        });
-
-    // `.foo` or `.foo(args)` in `atom.foo` or `atom.foo(args)`
-    let member_rhs = just(Token::Dot)
-        .ignore_then(field_name())
-        .then(method_call_rhs.or_not())
-        .map(|(method_or_field, method_call)| {
-            UnaryRhs::MemberAccess(UnaryRhsMemberAccess { method_or_field, method_call })
-        })
-        .labelled(ParsingRuleLabel::FieldAccess);
-
-    let just_a_dot =
-        just(Token::Dot).map(|_| UnaryRhs::JustADot).validate(|value, span, emit_error| {
-            emit_error(ParserError::with_reason(
-                ParserErrorReason::ExpectedIdentifierAfterDot,
-                span,
-            ));
-            value
-        });
-
-    let rhs = choice((call_rhs, array_rhs, cast_rhs, member_rhs, just_a_dot));
-
-    foldl_with_span(
-        atom(expr_parser, expr_no_constructors, statement, allow_constructors),
-        rhs,
-        |lhs, rhs, span| match rhs {
-            UnaryRhs::Call((is_macro, args)) => {
-                Expression::call(lhs, is_macro.is_some(), args, span)
-            }
-            UnaryRhs::ArrayIndex(index) => Expression::index(lhs, index, span),
-            UnaryRhs::Cast(r#type) => Expression::cast(lhs, r#type, span),
-            UnaryRhs::MemberAccess(field) => {
-                Expression::member_access_or_method_call(lhs, field, span)
+    fn eat_self(&mut self) -> bool {
```

Comment:
😲 🫢

Reply:
I do find it odd some `eat_` functions return a bool while other return Options. If it makes the call sites cleaner I think returning `Option<()>` would be nice even if IIRC we have to ignore a clippy warning for it.

Reply:
Like in the other comments, I'm not sure about this.